src/block-cache.ts

```diff
@@ -0,0 +1,183 @@
+import {
+  createAsyncMiddleware,
+  JsonRpcMiddleware,
+  JsonRpcRequest,
+} from 'json-rpc-engine';
+import * as cacheUtils from './cache-utils';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires,@typescript-eslint/no-require-imports
+const BlockTracker = require('eth-block-tracker');
+
+// `<nil>` comes from https://github.com/ethereum/go-ethereum/issues/16925
+const emptyValues: (string|null|undefined)[] = [undefined, null, '\u003cnil\u003e'];
+
+interface BlockCacheMiddlewareOptions{
+  blockTracker?: typeof BlockTracker;
+}
+
+export = createBlockCacheMiddleware;
+
+//
+// Cache Strategies
+//
+
+class BlockCacheStrategy {
+
+  private cache: Record<string, Record<string, unknown>>;
+
+  constructor() {
+    this.cache = {};
+  }
+
+  getBlockCacheForPayload(_payload: JsonRpcRequest<string[]>, blockNumberHex: string): Record<string, unknown> {
+    const blockNumber: number = Number.parseInt(blockNumberHex, 16);
+    let blockCache: Record<string, unknown> = this.cache[blockNumber];
+    // create new cache if necesary
+    if (!blockCache) {
+      const newCache: Record<string, unknown> = {};
+      this.cache[blockNumber] = newCache;
+      blockCache = newCache;
+    }
+    return blockCache;
+  }
+
+  async get(payload: JsonRpcRequest<string[]>, requestedBlockNumber: string): Promise<unknown> {
+    // lookup block cache
+    const blockCache: Record<string, unknown> = this.getBlockCacheForPayload(payload, requestedBlockNumber);
+    if (!blockCache) {
+      return undefined;
+    }
+    // lookup payload in block cache
+    const identifier: string|null = cacheUtils.cacheIdentifierForPayload(payload, true);
+    const cached: unknown = identifier ? blockCache[(identifier as string)] : undefined;
+    // may be undefined
+    return cached;
+  }
+
+  async set(payload: JsonRpcRequest<string[]>, requestedBlockNumber: string, result: Record<string, unknown>): Promise<void> {
+    // check if we can cached this result
+    const canCache: boolean = this.canCacheResult(payload, result);
+    if (!canCache) {
+      return;
+    }
+    // set the value in the cache
+    const blockCache: Record<string, unknown> = this.getBlockCacheForPayload(payload, requestedBlockNumber);
+    const identifier: string|null = cacheUtils.cacheIdentifierForPayload(payload, true);
+    blockCache[(identifier as string)] = result;
+  }
+
+  canCacheRequest(payload: JsonRpcRequest<string[]>,): boolean {
+    // check request method
+    if (!cacheUtils.canCache(payload)) {
+      return false;
+    }
+    // check blockTag
+    const blockTag: string = cacheUtils.blockTagForPayload(payload) as string;
+    if (blockTag === 'pending') {
+      return false;
+    }
+    // can be cached
+    return true;
+  }
+
+  canCacheResult(payload: JsonRpcRequest<string[]>, result: Record<string, unknown>): boolean {
+    // never cache empty values (e.g. undefined)
+    if (emptyValues.includes((result as unknown) as string)) {
+      return false;
+    }
```

Comment:
If the `result` can be something other than `Record<string, unknown>`, we should type it with an appropriate union type. After the `emptyValues.includes` type guard, TypeScript should be able to correctly infer that `result` is a `Record<string, unknown>`.